### ISSUE TITLE
Add support for Windows ARM64 as a target

### ISSUE THREAD

**Issue by stuartmorgan** [flutter/flutter#62597]
https://github.com/flutter/flutter/issues/53120 tracks supporting Windows arm64 as a development environment; this is for being able to build Flutter Windows applications that run on arm64.

This would require:
- Dart support for Windows arm64
- Dart support for cross-compiling Windows arm64 from x64 (since there are few Windows arm64 devices, requiring one for building would be problematic; for initial test purposes support without cross-compiling would be useful though)
- Engine build support for arm64
- CI and uploaded artifacts (again, this would require cross-compilation in order to be viable)
- Tooling support for managing arm64 builds, including fetching the artifacts, using separate build directories, etc.

**Comment by stuartmorgan** [comment]
Interesting, I didn't realize that was on the Dart team's radar.

I played around with this quite a while ago and got a PoC of the Dart part compiling, but then ran into the issue that it would require cross-building support, or for the entire compilation of the Flutter engine to be on aarm64 (bullet point #2). Does cross-compilation work at the Dart level (i.e., can Dart AOT compilation be done for arm64, but from an x64 machine)?

If not you could experiment with getting flutter/engine building on arm64, but getting from there to `flutter` tool support would require CI arm64 machines; I don't know if there's a good option for that available.

(/cc @cbracken )

**Comment by cbracken** [comment]
Aside from the compilation of the C++ bits, there's also gen_snapshot and the Dart compilation. Today we can generate a gen_snapshot binary that runs on x64 host and targets arm64, but it is not currently possible (without a significant amount of work from the Dart team) to build a gen_snapshot binary that runs on arm64 and targets x64 because there is no simx64 support in the Dart SDK.

On macOS, this isn't a huge problem since we can run x64 code on arm64 macs via Rosetta. It looks like this may [be possible](https://www.windowscentral.com/microsoft-states-x64-emulation-only-available-windows-11-arm-pcs) on Windows 11 and later. It looks like there was a preview release for Windows 10 based on this Windows [blog post](https://blogs.windows.com/windows-insider/2020/12/10/introducing-x64-emulation-in-preview-for-windows-10-on-arm-pcs-to-the-windows-insider-program/) but it's unclear if a final version will be available.

**Comment by stuartmorgan** [comment]
> Today we can generate a gen_snapshot binary that runs on x64 host and targets arm64

Nice! Given the current distribution of machines, that's the most important thing for us to have to get something working end-to-end.

**Comment by azchohfi** [comment]
@cbracken I believe we should look at these issues as separate issues.
This specific issue is tracking building Flutter apps that target Windows 10 and 11 on Arm64. Maybe we can create a cross compiling issue to track making a host Arm64 target x64 binaries.
Also, Windows 10 Arm64 will not support x64 emulation. That feature is exclusive to Windows 11 on Arm64, so we can't fully rely on the emulation to have coverage of Arm64. Energy consumption is also always in mind when any kind of emulation is used, so natively supporting Arm64 would provide a better battery life.

The way I see it, we have two options to unblock Arm64 Flutter apps on Windows 10/11:
1) Work to enable cross compilation on x64 hardware. If the compilation is on an Intel x64 machine, that would require cross compiling to Arm64 (which doesn't have a specific issue tracked by Flutter, but does have one from the Dart SDK, https://github.com/dart-lang/sdk/issues/28617, which would be a very first step in getting this to reality).
2) Work to enable all the tooling to work on an Arm64 host machine. If the compilation is happening on an Arm64 machine, then it would require work to enable **all** the flutter tools to works on Arm64 hosts, but there wouldn't be any cross-compilation effort required here.

Since cross compilation seems like a harder subject to tackle, and since Arm64 hardware on the cloud is starting to be a reality (e.g. Azure already has a [public preview](https://azure.microsoft.com/en-us/blog/now-in-preview-azure-virtual-machines-with-ampere-altra-armbased-processors/) of Arm64 VMs), working with option number 2 seems like a better approach. This would unblock developers to build their Flutter apps for Windows 10 Arm64 + Windows 11 Arm64, as well as enabling them to use Arm64 hosts to build their apps locally, with the same great developer tools that work today on Intel x64, with minimal focus on cross compilation.

Does that seem reasonable?

If so, would you be willing to help me drill this down a bit into actionable things to do? I would love to see this become a reality, specially since Microsoft announced [Project Volterra](https://blogs.windows.com/windowsdeveloper/2022/05/24/create-next-generation-experiences-at-scale-with-windows/).

**Comment by stuartmorgan** [comment]
> Also, Windows 10 Arm64 will not support x64 emulation. That feature is exclusive to Windows 11 on Arm64, so we can't fully rely on the emulation to have coverage of Arm64. Energy consumption is also always in mind when any kind of emulation is used, so natively supporting Arm64 would provide a better battery life.

`gen_snapshot` is only used during the build, so the emulation here (and associated restrictions) would only apply to developers at app creation time.

The first goal though would be to get the *engine* cross-compiling (x64->arm64), or compiling natively on arm64, or both. (And if only the later, a CI solution for creating engine artifacts). Without that we can't do anything at the tool level.

**Comment by cbracken** [comment]
Correct. The minimum set of steps required here (and I'm sure I'm missing a pile of items) is something like:
* [ ] Enable building x64 host binaries:
  * [ ] x64 build of gen_snapshot_arm64 (host arch: x64, target arch: arm64)
* [ ] Enable building arm64 target binaries for Windows:
  * [ ] flutter_engine.dll
  * [ ] flutter_windows.dll
  * [ ] libtessellator.dll (required?)
* [ ] Update flutter/recipes to build and upload new artifacts for use by the
      flutter tool. We should probably rename gen_snapshot to gen_snapshot_x64
      as part of this, but we'll need to preserve the old binary until the
      change rolls to the framework and the tool can be updated to use the
      binary with the new name.
* [ ] Update the gen_snapshot artifact resolution code in the tool to resolve
      the appropriate gen_snapshot depending on target architecture.
* [ ] Update the Windows app/plugin templates in the tool to support arm64.
* [ ] Update build framework CI with any updates necessary to build/execute
      these tests, benchmarks on arm64 devices.
* [ ] Create any CIPD packages required for arm64 Windows CI tests.
* [ ] Add arm64 Windows hardware to our infra pool.
* [ ] Add arm64 Windows images to our infra.
* [ ] Enable running integration tests, benchmarks on arm64 pool.

**Comment by azchohfi** [comment]
Yes, targeting an ARM64 host is probably an easier path forward. Visual Studio 2022 17.3 Preview 2+ runs as an ARM64 process on WIndows 11, and the toolchain should follow, so this is the least resistance path.
The Desktop development with C++ is exactly the type of app that is already supported, so Flutter fits perfectly here.
https://devblogs.microsoft.com/visualstudio/arm64-visual-studio/

**Comment by pbo-linaro** [comment]
Hello, I've been able to compile a native flutter app for arm64 (native build, no cross compilation). This is documented on [Linaro's Windows on Arm wiki](https://linaro.atlassian.net/wiki/spaces/WOAR/pages/28745138193/Dart+Flutter).

However, I needed to hack through flutter build scripts (see [details here](https://linaro.atlassian.net/wiki/spaces/WOAR/pages/28745138193/Dart+Flutter#%E2%9A%A0%EF%B8%8F-Current-status)). The goal is to upstream all that, starting first with dart-sdk.

Just wanted to say there does not seem to be any blocking issue for native build. Cross compilation will be next target.

**Comment by citrus77** [comment]
Great news!

**Comment by mohitmundhragithub** [comment]
From which version will we have the support ready?

**Comment by pbo-linaro** [comment]
I started to upstream bits to be able to build flutter engine natively under Windows on Arm (not yet targetting cross compilation)
Some more changes will be needed for third party repositories.

Once engine can be built easily, adding app support will be straightforward.

**Comment by pbo-linaro** [comment]
All patches needed were merged to be able to build flutter engine.
Last piece will be to add an option to flutter.bat to target specifically this platform. I'll make a PR for that.

**Comment by pbo-linaro** [comment]
I plan to add an option ```--target-platform``` to  ```flutter.bat build windows``` command(it already exists for ```build linux```). It will add custom flags to cmake call, allowing to build a native windows on arm app.

Does that sounds right to you?

**Comment by pbo-linaro** [comment]
I just created the PR to allow building natively arm64 flutter app 👍

**Comment by pbo-linaro** [comment]
For now, the PR adding arm64 native build is abandoned https://github.com/flutter/flutter/pull/113928. After discussion with @stuartmorgan (see PR), artifacts for this platform is a requirement before going further.

For the people expected this ASAP, you can follow those [build instructions](https://gitlab.com/Linaro/windowsonarm/packages/flutter/-/blob/master/recipe.sh) to get a native windows-arm64 app today. The patch it contains is what the PR was trying to upstream.

**Comment by pbo-linaro** [comment]
@stuartmorgan Windows on Arm can be deployed as a VM (for free) using one of those possibilities:
- KVM based virtual machine
- UTM based virtual machine (on apple M1/M2 hardware)
- Azure VM

I'll try to investigate cross-compilation, but meanwhile, could you contact people inside Google to know if it would be possible to get Windows on Arm VM. I sincerely hope a company with such resources is able to deploy a few VM somewhere on an arm64 server. That would allow users to get artifacts available now (flutter engine is building fine natively), while waiting to have cross compilation available later.

Thanks!

**Comment by loic-sharma** [comment]
I'm not very familiar with Flutter's builds, but these look like the relevant resources:

1. https://github.com/flutter/flutter/blob/master/dev/bots/README.md#editing-a-recipe
2. https://flutter.googlesource.com/recipes/+/refs/heads/main/recipes/engine/engine.py#1522

You likely won't have permissions to test your changes on Flutter's staging infrastructure using `led`. I've reached out internally to see what would be the best approach, but in the meantime, we can run `led` on your patches on your behalf. Feel free to ping us on the [Flutter desktop discord channel](https://discord.com/channels/608014603317936148/608020180177780791) for that.

FYI, we have holidays this week so our responses may be delayed! I will be out next week too.

**Comment by pbo-linaro** [comment]
Thanks @loic-sharma. I'll open a different issue (for engine artifacts) and try to make a PR on that recipe.

**Comment by vinc3m1** [comment]
Seems like this has stalled, any updates?

**Comment by dennisameling** [comment]
https://github.com/flutter/flutter/pull/176385 was merged back in March and made it into the 3.44.0 stable release 🎉 This means that the arm64 Dart SDK and Flutter Engine are now built for every new Flutter release.

When you [download the 3.44.0 SDK](https://storage.googleapis.com/flutter_infra_release/releases/stable/windows/flutter_windows_3.44.0-stable.zip) it still includes the x64 Dart SDK and Flutter Engine by default, but here's how you can force it to download the arm64 Dart SDK and Flutter Engine:

1. Download the 3.44.0 SDK to your Windows machine. Let's say we put it in `C:\flutter`.
2. Run `Remove-Item "C:\flutter\bin\cache\engine-dart-sdk.stamp" -Force`, otherwise `update_dart_sdk.ps1` won't notice it needs to download a fresh binary instead of using the stamped x64 version.
3. Run `dart --version`. This will automatically download the Windows arm64 version:

```
> dart --version
Checking Dart SDK version...
Downloading Dart SDK from Flutter engine ...
Expanding downloaded archive with PowerShell...
Building flutter tool...
Running pub upgrade...
Resolving dependencies...
Downloading packages...
Got dependencies.
Dart SDK version: 3.12.0 (stable) (Fri May 8 01:51:14 2026 -0700) on "windows_arm64"
```

4. Now run `flutter precache --windows`. This will download the arm64 engine:

```
> flutter precache --windows
[1/5] Flutter SDK
  ├─ [1/5] sky_engine                                            (1.5MB in 0.5s)
  ├─ [2/5] flutter_gpu                                           (0.0MB in 0.0s)
  ├─ [3/5] flutter_patched_sdk                                   (3.8MB in 0.9s)
  ├─ [4/5] flutter_patched_sdk_product                           (3.8MB in 0.9s)
  └─ [5/5] windows-arm64                                        (32.1MB in 8.0s)
[2/5] Windows SDK
  ├─ [1/4] windows-arm64-debug/windows-arm64-flutter           (77.5MB in 18.3s)
  ├─ [2/4] windows-arm64/flutter-cpp-client-wrapper              (0.1MB in 0.0s)
  ├─ [3/4] windows-arm64-profile/windows-arm64-flutter         (65.0MB in 15.7s)
  └─ [4/4] windows-arm64-release/windows-arm64-flutter         (63.1MB in 15.1s)
[5/5] windows-arm64/font-subset                                  (2.0MB in 0.5s)
```

## What ships inside the ZIPs (identical arch coverage)

| | 3.41.9 | 3.44.0 |
|---|---|---|
| Bundled Dart SDK | x64, v3.11.5 | x64, v3.12.0 |
| Bundled engine artifacts | `windows-x64`, `-profile`, `-release` only | `windows-x64`, `-profile`, `-release` only |
| Engine revision (`bin/cache/engine.stamp`) | `42d3d75a56efe1a2e9902f52dc8006099c45d937` | `4c525dac5ebe5971c5708ef73558ed8edcf4a362` |
| `bin/internal/update_dart_sdk.ps1` arm64 logic | present (identical) | present (identical) |

Both archives bundle **only** an x64 Dart + x64 engine, and ship a **byte-identical** `update_dart_sdk.ps1` that already knows how to fetch `dart-sdk-windows-arm64.zip` when `PROCESSOR_ARCHITECTURE=ARM64`. So the tooling is not the gate; availability in the bucket is.

## Bucket availability, keyed by engine revision

Here we can see that 3.44.0 really is the first version that ships the necessary arm64 binaries.

Base URL: `https://storage.googleapis.com/flutter_infra_release/flutter/<engineRevision>/<path>`

| Artifact (`<path>`) | 3.41.9 (`42d3d75a…`) | 3.44.0 (`4c525dac…`) |
|---|---|---|
| `dart-sdk-windows-arm64.zip` | **404** | **200** |
| `windows-arm64/artifacts.zip` | **404** | **200** |
| `windows-arm64/windows-arm64-embedder.zip` | 404 | **200** |
| `windows-arm64/flutter-cpp-client-wrapper.zip` | 404 | **200** |
| `windows-arm64/impeller_sdk.zip` | 404 | **200** |
| `windows-arm64/font-subset.zip` | 404 | **200** |
| `windows-arm64-release/windows-arm64-flutter.zip` | **404** | **200** |
| `windows-arm64-profile/windows-arm64-flutter.zip` | 404 | **200** |
| `windows-arm64-debug/windows-arm64-flutter.zip` | 404 | **200** |
| `dart-sdk-windows-x64.zip` (control) | 200 | 200 |
| `windows-x64/artifacts.zip` (control) | 200 | 200 |

**Comment by kjxbyz** [comment]
So, when will I be able to package Windows applications that support the ARM64 architecture on a Windows ARM64 machine?

**Comment by dennisameling** [comment]
> So, when will I be able to package Windows applications that support the ARM64 architecture on a Windows ARM64 machine?

@kjxbyz you should be able to do so with the instructions above. As of 3.44.0, you can switch between the Dart x64 and arm64 SDK on Windows. It still defaults to x64 though, so you need to explicitly pull in the arm64 version. The latter will build arm64 binaries by default. More details are here: https://github.com/flutter/flutter/issues/186730#issuecomment-4573214964

**Comment by kjxbyz** [comment]
Microsoft, Qualcomm, Nvidia, and OEMs have invested heavily in Windows on ARM, hoping for significant progress in this feature.
